### PR TITLE
Avx512 memcpy small

### PR DESCRIPTION
**Pull request by yaqi-zhao**
### Changelog category (leave one):

- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add x86 avx512 support for memcpySmall functions to accelerate memory copy. It works only if you 
compile ClickHouse by yourself.

Detailed description / Documentation draft:
The changes is to implement the memCpy function in memcpySmall.h by using avx512 intrinsic '_mm512_storeu_si512'
and '_mm512_loadu_si512' when copied memory exceed 64bytes. Compared with the original version, functions in the avx512
version can have better performance(memcpySmallAllowReadWriteOverflow15Impl) or the same performance.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/yaqizhao/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/yaqizhao/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Case(byte) | memcpySmallAllowReadWriteOverflow15(SSE2) | memcpySmallAllowReadWriteOverflow15(AVX512)
-- | -- | --
8 | 20 | 20
16 | 27 | 26
32 | 30 | 24
64 | 39 | 33
128 | 49 | 37
256 | 66 | 44
512 | 75 | 70

</body>

</html>

Test Case:
Q2: SELECT C_NATION,S_NATION,toYear(LO_ORDERDATE) AS year,sum(LO_REVENUE) AS revenue FROM lineorder_flat WHERE C_REGION = 'ASIA' AND S_REGION = 'ASIA' AND year >= 1992 AND year <= 1997 GROUP BY C_NATION, S_NATION, year ORDER BY year ASC, revenue DESC;
Q3:  SELECT toFixedString('foo', 128); SELECT toFixedString('foo', 64); SELECT toFixedString('foo', 32); SELECT toFixedString('foo', 16);
Q4:SELECT toFixedString('foo', 128);
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/yaqizhao/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/yaqizhao/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">

Query | Baseline  (QPS) | avx512 optimized (QPS) | improvement
-- | -- | -- | --
Q2 | 109 | 112 | ~
Q3 | 2635 | 2700 | 2.80%
Q4 | 1910 | 2024 | 5.90%

</body>

</html>

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by kitaisreal** [comment]
@Mergifyio update

File: src/Common/memcpySmall.h [hunk 1]
```diff
@@ -27,18 +23,58 @@
...
 namespace detail
 {
     inline void memcpySmallAllowReadWriteOverflow15Impl(char * __restrict dst, const char * __restrict src, ssize_t n)
     {
         while (n > 0)
-        {
+        { 
+          if (n >= 64) {
+            _mm512_storeu_si512(reinterpret_cast<__m512i *>(dst),
+            _mm512_loadu_si512(reinterpret_cast<const __m512i *>(src)));
```
Comment (yaqi-zhao) [inline src/Common/memcpySmall.h:39]: comment

File: src/Common/memcpySmall.h [hunk 2]
```diff
@@ -27,18 +23,62 @@
   * Use with caution.
   */
 
+#if defined(__AVX512F__) && !defined(MEMORY_SANITIZER)
```
Comment (alexey-milovidov) [inline src/Common/memcpySmall.h:26]: Missing comment, why `!defined(MEMORY_SANITIZER)` here?
And why it is not required below, in `#elif defined(__SSE2__)`?

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]
I changed code formatting,
more changes required.

**Comment by alexey-milovidov** [comment]
The improvement is only marginal.

But there are the following caveats:
1. This code path will be run only if ClickHouse is compiled with AVX-512 globally enabled. But we don't do that, because AVX-512 is rare. And we don't support any builds of ClickHouse except the single official build.
2. Implementing dynamic dispatching for this code will be difficult if not impossible.
3. The changes have been tested on unknown CPU model, but the performance of AVX-512 can vary a lot on different CPU models and on some models it can lead to performance degradation.
4. We add 15 byte padding to the columns, but with the introduction of AVX-512, it will be reasonable to change the padding to 63 bytes and simplify the code even more. It was not done in this PR.
5. It has been tested on TPC-H or Star Schema Benchmark dataset that is known to represent unrealistic workload with too many pseudorandom data with unnatural distributions. It may lead to excessive copying and overly-optimistic results on this benchmark.
6. Some prior work https://github.com/ClickHouse/ClickHouse/pull/21812 and https://presentations.clickhouse.com/zero_cost_conf/memcpy/ advices against this optimization.

**Comment by alexey-milovidov** [comment]
Please resubmit if we will come up to some other solution.

**Comment by yaqi-zhao** [comment]
Thanks for your comments. @alexey-milovidov  It is useful and helps us provide more comprehensive solution in the future. But there are some confusions about the above caveats;
1.	I found that prior work on avx512 optimization have been merged, e.g.,,Avx512 memcmp small [#33706](https://github.com/ClickHouse/ClickHouse/pull/33706) && add x86 feature avx2/avx512 support for accelerating filter operation [#30014](https://github.com/ClickHouse/ClickHouse/pull/30014). But the comment said ck don't support any builds of ClickHouse except the single official build. I have doubt that if we have more avx512 optimization in the future, can these features be merged?
2.	Comments say “The improvement is only marginal” --=>  Is that means 5% performance improvement on function query tested by Star Schema Benchmark to small? Generally, how much the improvement made will the PR be accepted?

**Comment by alexey-milovidov** [comment]
5% is good if it is available in the default build configuration and not good to motivate some code path that works only in custom builds.

**Comment by alexey-milovidov** [comment]
I should better invite @nickitat here instead of closing :)

**Comment by yaqi-zhao** [comment]
@alexey-milovidov  what if I add a cmake option to enable the code by default if build enviromnent supports AVX512?

**Comment by alexey-milovidov** [comment]
It's not a good option, we can build on servers with AVX-512 but deploy on other servers.
Dynamic dispatch will be better if they can really work on memcpySmall (that's in question).

**Pull request closed without merging**